compiler/rustc_infer/src/infer/outlives/components.rs

```diff
@@ -71,6 +71,26 @@ fn compute_components<'tcx>(
     // in the `subtys` iterator (e.g., when encountering a
     // projection).
     match *ty.kind() {
+            // HACK(oli-obk): ignore lifetimes found in opaque types.
+            // See https://github.com/rust-lang/rust/issues/84305 for
+            // why registering the lifetimes can actually be unsound.
+            ty::Opaque(_, _) => for ty in ty.walk() {
+                match ty.unpack() {
+                    GenericArgKind::Type(ty) => {
+                        match ty.kind() {
+                            // ???! needed for async blocks
+                            ty::Closure(..) | ty::Generator(..) | ty::Projection(_) => {
+                                compute_components(tcx, ty, out, visited);
+                            }
+                            _ => (),
+                        }
+                    }
+                    GenericArgKind::Lifetime(_) => {}
+                    GenericArgKind::Const(_) => {
+                        compute_components_recursive(tcx, ty, out, visited);
+                    }
+                }
+            },
```

Comment:
I suggest we approach this the same way we do for projections:

Push a "opaque-component" and then skip over the contents in some cases, expand in others. I can supply more links if desired. :)